subprojects/core/src/test/groovy/org/gradle/internal/classpath/GroovyDynamicDispatchingInterceptingTest.groovy

```diff
@@ -138,6 +138,35 @@ class GroovyDynamicDispatchingInterceptingTest extends AbstractCallInterceptionT
         modifiedMetaClasses.every()
     }
 
+    def 'closure does not erroneously become effectively instrumented after it threw an exception'() {
+        given:
+        def transformedThrowingClosure = instrumentedClasses.instrumentedClosure {
+            throw new RuntimeException()
+        }
+        transformedThrowingClosure.delegate = new InterceptorTestReceiver()
+        def transformedInterceptedClosure = instrumentedClasses.instrumentedClosure {
+            test()
+        }
+        transformedInterceptedClosure.delegate = new FalseInterceptorTestReceiver()
+
+        when: 'a closure throwing an exception is called, then a closure that hits an instrumented call is called'
+        try {
+            transformedThrowingClosure()
+            // this calls throws an exception, but we expect the closure to be still correctly removed from control flow tracking
+        } catch (RuntimeException ignored) {}
```

Comment:
❔  Could we have used Spock's `thrown()` helper here to assert that an exception was throw?

I would then introduce a second `when:` clause for the `transformedInterceptedClosure()` call.

Reply:
Done, thanks!